### PR TITLE
GCP one click deployment

### PR DESCRIPTION
**Pull request by EugeneLightsOn**
"deployment: add GCP Cloud Run Button deployment"

This PR makes changes to the Dockerfile and adds a new Google Cloud Platform (GCP) deployment guide. 

## Dockerfile
- The base image has been changed from `buildpack-deps:buster` to `python:3.11`.
- The `FROM` instruction now specifies `ghcr.io/cohere-ai/terrarium:latest as terrarium`.
- The `ENV` instructions have been updated to set the `PG_APP_HOME` and `PYTHONPATH` variables.
- The `WORKDIR` instruction has been updated to set the working directory to `/workspace/src/`.
- The `RUN` instruction has been modified to install `nginx` and `nodejs` using `apt-get`.
- The `COPY` instruction has been added to copy the `gcp-entrypoint.sh` script to `/sbin/`.
- The `RUN` instruction has been modified to install `poetry` with the option `--without setup`.
- The `EXPOSE` instruction has been updated to expose ports `4000/tcp`, `8000/tcp`, and `8090/tcp`.
- The `CMD` instruction has been updated to specify the entrypoint script as `gcp-entrypoint.sh`.

## New Files
- `gcp-entrypoint.sh`: This script sets up the runtime environment and launches the necessary applications.
- `nginx.conf`: This file contains the configuration for the Nginx web server.

## app.json
- The `name` field has been changed from `"one-click-demo"` to `"toolkit-deploy"`.
- The `DATABASE_URI` field has been replaced with `DATABASE_URL`, which now includes the description "URL for your Postgres database".
- A new `COHERE_API_KEY` field has been added to the `env` object, which includes the description "Your Cohere API key".
- A new `hooks` object has been added with a `prebuild` command to configure Docker authentication using `gcloud`.
- The `options` object has been updated to include the `memory`, `cpu`, `port`, and `max-instances` fields.

## docker_scripts/env-defaults
- The `PYTHON_INTERPRETER_URL` export has been updated to use `localhost:8080` as the default value.

## docker_scripts/functions
- A new `run_nginx` function has been added to start the Nginx proxy.
- The `run_backend` function has been updated to include the `DATABASE_URL` variable.
- A new `run_backend_poetry` function has been added to handle the backend installation and migration.
- A new `run_frontend_proxy` function has been added to start the frontend proxy.

## docs/deployment_guides/gcp_one_click_deployment.md
- A new guide has been added to outline the steps for deploying the Toolkit to GCP using the "One Click" deployment option.
- It includes instructions on setting up a PostgreSQL database, configuring the `DATABASE_URL`, providing the Cohere API key, and allowing access to git registries.
- The guide also covers post-deployment steps, such as accessing the Toolkit URL and handling possible deployment errors.

## docs/service_deployments.md
- A new link has been added to the "One Click Deploy to GCP" guide in the `deployment_guides` directory.

### CODE REVIEW

File: docs/deployment_guides/gcp_one_click_deployment.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# Google Cloud Platform One Click Deployment Guide
+
+Before deploying to Google Cloud Run, you'll need a postgres database accessible to your Google Cloud Project, authenticated by a username and password. You'll be prompted for a `DATABASE_URL` before the container builds.
+
+[![Run on Google Cloud](https://deploy.cloud.run/button.svg)](https://deploy.cloud.run)
+
+## Deployment Steps
+After clicking the button above, you'll be taken to the Google Cloud Console. Follow these steps to deploy the Toolkit:
+- **Select a Project**: Choose the Google Cloud Project you want to deploy the Toolkit to.
```
Comment (tianjing-li) [inline docs/deployment_guides/gcp_one_click_deployment.md:9]: newline before

File: docs/deployment_guides/gcp_one_click_deployment.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+## Post-Deployment Steps
+After the deployment process is complete, you'll be able to access the Toolkit by navigating to the URL provided in the Google Cloud Console.
+To get this URL, follow these steps:
+- Click on the "Cloud Run" button on the left side of the screen
+- Click on the toolkit-deploy service
+- Click on the "URL" link to open the Toolkit in a new tab
+
+## Possible Deployment Errors
+If the deployment is stuck on pushing the image to the Container Registry close the shell and try to redeploy the service.
+If the deployment is stuck in a pending state or the connection to your Google Cloud Shell was lost like it shown on the image below:
```
Comment (tianjing-li) [inline docs/deployment_guides/gcp_one_click_deployment.md:24]: Nit: newline between each sentence here

**Review by tianjing-li — APPROVED** [review]
This is great - some very small asks

File: docs/deployment_guides/gcp_one_click_deployment.md (hunk 2 above)
Comment (EugeneLightsOn) [reply docs/deployment_guides/gcp_one_click_deployment.md:28]: Ty

File: docs/deployment_guides/gcp_one_click_deployment.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+Optionally you can get this URL using GCP console, to do it follow these steps:
+- Click on the "Cloud Run" button on the left side of the screen
+- Click on the toolkit-deploy service
+- Click on the "URL" link to open the Toolkit in a new tab
+
+## Possible Deployment Errors
+
+If the deployment is stuck on pushing the image to the Container Registry close the shell and try to redeploy the service.
+
+If the deployment is stuck in a pending state or the connection to your Google Cloud Shell was lost like it shown on the image below:
```
Comment (tianjing-li) [inline docs/deployment_guides/gcp_one_click_deployment.md:28]: lost like shown below:

File: docs/deployment_guides/gcp_one_click_deployment.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+![](/docs/assets/cloud_shell_stuck.png)
+
+or you get the nginx 502 error when you try to access the Toolkit URL,
```
Comment (tianjing-li) [inline docs/deployment_guides/gcp_one_click_deployment.md:31]: when trying to access the Toolkit URL

**Review by tianjing-li — APPROVED** [review]
Approved- minor comments

File: docs/deployment_guides/gcp_one_click_deployment.md (hunk 3 above)
Comment (EugeneLightsOn) [reply docs/deployment_guides/gcp_one_click_deployment.md]: Ty, TJ

File: docs/deployment_guides/gcp_one_click_deployment.md (hunk 4 above)
Comment (EugeneLightsOn) [reply docs/deployment_guides/gcp_one_click_deployment.md]: Thanks

**Review by tianjing-li — APPROVED** [review]

**Pull request merged**